src/desktop/window.rs

```diff
@@ -374,6 +374,57 @@ impl Window {
         found.into_inner()
     }
 
+    /// Damage of all the surfaces of this window
+    pub(super) fn accumulated_damage(&self) -> Vec<Rectangle<i32, Logical>> {
+        let mut damage = Vec::new();
+        let location = (0, 0).into();
+        if let Some(surface) = self.0.toplevel.get_surface() {
+            with_surface_tree_upward(
+                surface,
+                location,
+                |_surface, states, location| {
+                    let mut location = *location;
+                    if let Some(data) = states.data_map.get::<RefCell<SurfaceState>>() {
+                        let data = data.borrow();
+                        if data.texture.is_none() {
+                            if states.role == Some("subsurface") {
+                                let current = states.cached_state.current::<SubsurfaceCachedState>();
+                                location += current.location;
+                            }
+                            return TraversalAction::DoChildren(location);
+                        }
+                    }
+                    TraversalAction::SkipChildren
+                },
+                |_surface, states, location| {
+                    let mut location = *location;
+                    if let Some(data) = states.data_map.get::<RefCell<SurfaceState>>() {
+                        let data = data.borrow();
+                        let attributes = states.cached_state.current::<SurfaceAttributes>();
+
+                        if data.texture.is_none() {
+                            if states.role == Some("subsurface") {
+                                let current = states.cached_state.current::<SubsurfaceCachedState>();
+                                location += current.location;
+                            }
+
+                            damage.extend(attributes.damage.iter().map(|dmg| {
+                                let mut rect = match dmg {
+                                    Damage::Buffer(rect) => rect.to_logical(attributes.buffer_scale),
+                                    Damage::Surface(rect) => *rect,
+                                };
+                                rect.loc += location;
+                                rect
+                            }));
```

Comment:
On the longer-term, this could take into account the `opaque_region` specified by the client to discard damage that is completely behind an opaque surface.

To transpose that on the space level, the Window would need to have a method returning a `Vec<Region>` describing which parts of it are opaque I guess.

Reply:
Yeah definitely. There is definitely room for improvement, but this is getting us probably already 80% of the way in terms of performance. (I still need to actually benchmark/profile it.)

Reply:
Lets make another ticket for this one.